### PR TITLE
fix: first round of bugfixes for v4

### LINKED ISSUE #304: [v4][no-await-sync-query] False positives on appearance/disappearance examples
In https://testing-library.com/docs/guide-disappearance, there's some techniques outlined for waiting for elements to be removed:

```js
test('movie title appears', async () => {
  // element is initially not present...

  // wait for appearance inside an assertion
  await waitFor(() => {
    expect(getByText('the lion king')).toBeInTheDocument()
  })
})
```

```
test('movie title no longer present in DOM', async () => {
  // element is removed
  await waitForElementToBeRemoved(() => queryByText('the mummy'))
})
```

Both of these examples are flagged by `no-await-sync-query`, and they probably shouldn't be.

### LINKED ISSUE #307: First round of v4 beta issues
The first round of issues I found while testing v4 beta:

- problem A: `await-async-utils` doesn't report the right node name
  CleanShot 2021-03-28 at 20 54 18@2x
- problem B: `prefer-find-by` is reporting an incorrect message
  CleanShot 2021-03-28 at 21 24 41@2x
- problem C: `no-await-sync-query` reports an error when a parent is awaited (related to #304 )
  CleanShot 2021-03-28 at 20 55 14@2x
- problem D: `no-wait-for-multiple-assertions` should display the error on each assertion other than the first one instead of a single error on `waitFor`
- problem E: `no-wait-for-side-effects` should display the error on each side effect instead of a single error on `waitFor`
- ~~problem F: ESLint `peerDependency` seems set incorrectly~~ this is actually fine, it seems a warning due to installing it through `resolutions` and ESLint being a sub-dependency of `react-scripts`
  CleanShot 2021-03-29 at 21 54 25@2x

### PR DESCRIPTION
**Pull request by Belco90**
Closes #307. Closes #304

### CODE REVIEW

File: lib/detect-testing-library-utils.ts [hunk 1]
```diff
@@ -335,12 +335,20 @@ export function detectTestingLibraryUtils<
...
+
+      if (userEvent) {
+        userEventName = userEvent.name;
+      } else if (isAggressiveModuleReportingEnabled()) {
+        userEventName = USER_EVENT_NAME;
+      }
+
+      if (!userEventName) {
+        return false;
+      }
```
Comment (Belco90) [inline lib/detect-testing-library-utils.ts:349]: Same issue I had with `isUserEventMethod`: `userEvent` util won't be imported from Testing Library module and it's a default import, so it has to be checked against its own import specifier.

File: lib/node-utils.ts [hunk 2]
```diff
@@ -153,19 +153,6 @@ export function findClosestCallNode(
   }
 }
 
-export function isCallExpressionCallee(
```
Comment (Belco90) [inline lib/node-utils.ts:156]: Not used anymore. It was a little bit useless, to be honest.

File: lib/rules/await-async-utils.ts [hunk 3]
```diff
@@ -80,7 +80,7 @@ export default createTestingLibraryRule<Options, MessageIds>({
                   node,
                   messageId: 'awaitAsyncUtil',
                   data: {
-                    name: referenceNode.name,
+                    name: node.name,
```
Comment (Belco90) [inline lib/rules/await-async-utils.ts:83]: This fixes problem A from #307

File: lib/rules/no-await-sync-query.ts [hunk 4]
```diff
@@ -29,22 +26,19 @@ export default createTestingLibraryRule<Options, MessageIds>({
...
 
-        if (helpers.isSyncQuery(node)) {
+        if (helpers.isSyncQuery(deepestIdentifierNode)) {
           context.report({
-            node,
+            node: deepestIdentifierNode,
             messageId: 'noAwaitSyncQuery',
             data: {
-              name: node.name,
+              name: deepestIdentifierNode.name,
```
Comment (Belco90) [inline lib/rules/no-await-sync-query.ts:41]: This fixes problem C from #307

File: lib/rules/no-wait-for-multiple-assertions.ts [hunk 5]
```diff
@@ -46,14 +54,20 @@ export default createTestingLibraryRule<Options, MessageIds>({
...
         return;
       }
 
-      context.report({
-        node: callExpressionNode,
-        messageId: 'noWaitForMultipleAssertion',
-      });
+      for (let i = 0; i < expectNodes.length; i++) {
+        if (i !== 0) {
+          context.report({
```
Comment (Belco90) [inline lib/rules/no-wait-for-multiple-assertions.ts:65]: This fixes problem D from #307

File: lib/rules/no-wait-for-side-effects.ts [hunk 6]
```diff
@@ -49,14 +57,17 @@ export default createTestingLibraryRule<Options, MessageIds>({
...
+      const sideEffectNodes = getSideEffectNodes(node.body);
+      if (sideEffectNodes.length === 0) {
         return;
       }
 
-      context.report({
-        node: callExpressionNode,
-        messageId: 'noSideEffectsWaitFor',
-      });
+      for (const sideEffectNode of sideEffectNodes) {
```
Comment (Belco90) [inline lib/rules/no-wait-for-side-effects.ts:65]: This fixes problem E from #307

File: lib/rules/prefer-find-by.ts [hunk 7]
```diff
@@ -53,13 +53,13 @@ export default createTestingLibraryRule<Options, MessageIds>({
...
       description:
-        'Suggest using find* instead of waitFor to wait for elements',
+        'Suggest using `find*` query instead of `waitFor` + `get*` to wait for elements',
       category: 'Best Practices',
       recommended: 'warn',
     },
     messages: {
       preferFindBy:
-        'Prefer {{queryVariant}}{{queryMethod}} method over using await {{fullQuery}}',
+        'Prefer `{{queryVariant}}{{queryMethod}}` query over using `{{waitForMethodName}}` + `{{prevQuery}}`',
```
Comment (Belco90) [inline lib/rules/prefer-find-by.ts:62]: This fixes problem B from #307

**Review by timdeschryver — APPROVED** [review]

**Pull request merged**